### ISSUE TITLE
feature request : sort parameter name in alphabetical order in the device explorer

### ISSUE THREAD

**Issue by avilleret** [ossia/score#344]
it would be more convenient to my eyes to see parameters listed in alphabetical order in the device explorer rather

**Comment by bltzr** [comment]
rather than ?
currently (at least when queried from the Jamoma implementation of Minuit) they are sorted based on the priority attribute, which is very important: so parameters are stored, and then recalled in the right order
what could be done, though, is that: 
- all nodes with a priority attribute are sorted based on the priority value
- all other nodes are sorted alphabeticall, *after* the prioritized nodes
what do you @OSSIA/users think about this ?

**Comment by avilleret** [comment]
...rather than in kind of random order :
![image](https://cloud.githubusercontent.com/assets/504914/21278363/0598332a-c3db-11e6-9739-9441f229d00a.png)

What about a clickable column header that define the attribute to sort and the direction ?

**Comment by bltzr** [comment]
> What about a clickable column header that define the attribute to sort and the direction ?

Sure, but then again, I think we should have a default behaviour (and the one I proposed has been agreed on collectively by the Jamoma collective, and I think it has some good sides)

**Comment by jcelerier** [comment]
(I should just emphasize that as of today the priority attribute isn't handled either... doing this asap)

**Comment by avilleret** [comment]
@bltzr I agree with the default behavior you propose

**Comment by bltzr** [comment]
> (I should just emphasize that as of today the priority attribute isn't handled either... doing this asap)

AFAIK Jamoma replies to i-score's namespace queries in this way, and then i-score just keeps the received order - maybe ossia.device should just do the same ? At least for now ?

Because there could be a potential problem for other protocols, e.g. OSC: sorting alphabetically could scramble a carefully ordered namespace (made e.g. with a namespace file), which doesn't necessary number each and every parameter with a priority... just trying to anticipate problems here...

**Comment by jcelerier** [comment]
There is something that I don't quite understand : why should the way it is displayed be related to the priority (or whatever network feature ?) ?

**Comment by jln-** [comment]
Exactly what I was wondering as well. The fact that messages are sent/received following defined priorities is probably not to be debated of.

However having addresses in the devices explorer sorted alphabetically seems much more convenient/user friendly to me. I guess I'd be fine if priorities are dealt with internally without me having to bother about it.

**Comment by bltzr** [comment]
well, in previous i-score versions, there wasn't a separation between the explorer (where the tree structure is shown) and the inspector (where the state contents are displayed), so I guess it was a simpler solution to have everything following the same logic (sorted by priorities)

but, now, things are slightly different, so I guess we could change this in the explorer - even though I'm not sure displaying nodes alphabetically is always the best solution: sometimes you want to use a different logic to sort your devices' trees... using priorities for both visual ordering and priority management seemed to be a good solution, but I guess it's debatable, and I suppose there might be better solutions now that we have a full-fledged developer at hand ;-)

concerning the state content editors, though, I still think it's important to relate the visual order and the order in which messages are sent when recalling the state - don't you think so, too ?

so... maybe we could have a default behavior, as proposed (or otherwise, if there are strong arguments), and then we could leave the possibility to the user to move the order of things, and also to sort the nodes based on whichever criteria they want...

I don't know how urgent this all is, though... not very much, I'd say...

And probably for now, we should manage the initial ordering from the device side, and at least for now let i-score display nodes in the order where they were received (so we don't break existing behaviors)

**Comment by jln-** [comment]
> even though I'm not sure displaying nodes alphabetically is always the best solution: sometimes you want to use a different logic to sort your devices' trees

Yes, other sorting/filtering options would be useful, for sure. I meant alphabetically sorting seems better or user friendly as a default option.

> concerning the state content editors, though, I still think it's important to relate the visual order and the order in which messages are sent when recalling the state - don't you think so, too ?

Yes, I agree.

> I don't know how urgent this all is, though... not very much, I'd say...

Not ultra urgent. This is amongst the things we can evaluate during the workshops

**Comment by bltzr** [comment]
there's a little bit of discussion under https://github.com/OSSIA/i-score/issues/344 - what seems to be important is that state messages are actually recalled (hence stored) by order of ascending priority, with the exception of priority 0 (or no priority) recalled last

**Comment by bltzr** [comment]
@avilleret proposed that the higher the number the higher the priority - this definitely makes sense so I propose that we adopt this (which also doesn't require the somehow strange priority=0 exception)

**Comment by jln-** [comment]
I was never really fond of the way we did it in Jamoma. 

Rather than switching the order of the numbering, I'd rather rename the priority attribute as 'order'. Otherwise, I bet set something like 'priority 100' just in case for parameters I need to be recalled first, then 99, for second one etc. Which is not very intuitive either.

'at Order' 1, 2, 3 and so on seems both convenient and explicit. Dont you think ?

FWIW, after checking this out, order 1, 2, 3 is how it is made in Max4Live so this kind of comforts my impression this would work better.

**Comment by bltzr** [comment]
actually, after switching my mind a little, I actually find it's much better to have the numbers reversed - among others because then we don't have to make this 0 exception
I find @avilleret 's phrasing in ossia.param-help.pd quite clear and unambiguous: 
"@priority <int> defines recall order, higher priority comes first"

**Comment by jln-** [comment]
@avilleret documentation is clear and concise. No problem here. But having something's documentation clear, doesn't mean this something is. :-)

I agree that '_at_priority [+∞ / 0]' is semantically better but this doesn't totally solve the inherent ambiguity. Problem here is not so much the numbers direction but rather the priority very word that implies this ranking approach where we want to control a sequence order.

_at_priority [1-99] (as it is in Jamoma or Max'pattr things) mixes both approaches which is bad imho. Agree on that.

Still, switching the numbering order leaves a kind decorrelation between how we describe what is happening: '_at_priority +∞' will get recalled 1st, while '_at_priority 1' will get recalled last. Sounds like speaking backward to me (and I hardly find a situation where this would be more intuitive)...

In that regard, '_at_order [1-+∞]' looks as unambiguous as possible to a point where it's almost self explanatory.

Last but not least, in Max only there are two ways to address this sequencing recalling order issue, so I'd rather stick to one convention rather than adding a new one. But sure, I'd rather stick to the non broken one. :-)

We may want to check in other environment how this is addressed ?

**Comment by bltzr** [comment]
I’m not sure I fully understand
in your proposal which is output first ?
_at_order 1 or _at_order 2 ?

**Comment by jcelerier** [comment]
> '_at_priority +∞' will get recalled 1st, while '_at_priority 1' will get recalled last.

I really don't understand how this is a problem. "Maximum priority", "Highest priority" is I think quite common vocabulary. If you go to the supermarket, the people who have highest priority are the one who go first through the cashier too and this does not seem to cause misunderstandings. Likewise when driving.

**Comment by jln-** [comment]
Sorry for going to fast. The point I see with using 'order' is that what you set is what you get: so _at_order 1 is output first, _at_order 2 is output second, etc.

0 == no order defined. This may lead to an exception in the code (?) but user-wise I find it quite intuitive as well.

**Comment by jcelerier** [comment]
> 0 == no order defined. 

why not just "not setting" the order in order to have it not being defined ?

**Comment by jln-** [comment]
> I really don't understand how this is a problem. "Maximum priority", "Highest priority" is I think quite common vocabulary. If you go to the supermarket, the people who have highest priority are the one who go first through the cashier too and this does not seem to cause misunderstandings. Likewise when driving.

Very good point: we dont want a software to look like a supermarket :-)

Joke aside, sure this is perfectly clear when expressed as words. Possible confusion comes from translating those words into numbers. I bet there are number of examples everywhere where "highest" will be 1 in one context or 100 in another.

If I may try a metaphor, the one on the _highest_ step of a podium is the one who arrived 1 at a race. The one on the lowest step is the one who arrived in 3 position.

Sure we do the mental conversion, but I'm just wondering wouldn't be straightforward to just set the position of arrival rather than the height of the step in the podium.

> why not just "not setting" the order in order to have it not being defined ?

Sure sure. Just meant to say we count from 1 - 0 (or whatever negative value) being a default value for the attribute

**Comment by bltzr** [comment]
@jln- maybe you're too used to the way pattr works in Max (and then Jamoma, because it mimicks its...). I had the same reaction, first.
habit set aside, I think this proposal is perfectly clear and understandable (Higher numbers mean higher priorities), and somehow less convoluted (because of the 0 thingie...) than the pattr proposal.
